aineko/extras/connectors/websocket_client.py

```diff
@@ -0,0 +1,216 @@
+    then replace `{$SECRET_NAME}` with `SECRET_VALUE` before connecting to the
+    WebSocket.
+
+    Example usage in pipeline.yml:
+    ```yaml title="pipeline.yml"
+    pipeline:
+      nodes:
+        WebSocketClient:
+          class: aineko.extras.WebSocketClient
+          outputs:
+            - test_websocket
+          node_params:
+            url: "wss://example.com"
+            header:
+              auth: "Bearer {$SECRET_NAME}"
+            init_messages:
+                - {"Greeting": "Hello, world!"}
+    ```
+    """
+
+    retry_count = 0
+
+    def _pre_loop_hook(self, params: dict | None = None) -> None:
+        """Initialize the WebSocket connection.
+
+        Raises:
+            ValueError: If the params are invalid.
+        """
+        # Cast params to ParamsWebSocketClient type
+        try:
+            if params is not None:
+                self.ws_params = ParamsWebSocketClient(**params)
+            else:
+                raise ValueError(
+                    "No params provided to WebSocket connector. "
+                    "Node requires at least a url param."
+                )
+        except Exception as err:  # pylint: disable=broad-except
+            # Cast pydantic validation error to ValueError
+            # so that it can be properly caught by the node
+            # Note: this is required because pydantic errors
+            # are not picklable
+            raise ValueError(
+                "Failed to cast params to ParamsWebSocketClient type. "
+                f"The following error occurred: {err}"
+            ) from err
+
+        # Create the websocket subscription
+        self.ws = websocket.WebSocket()
+        self.create_subscription()
+
+    def _execute(self, params: dict | None = None) -> None:
+        """Polls and gets data from the WebSocket.
+
+        Raises:
+            ValueError: If the retry count exceeds the max retries.
+        """
+        try:
+            # Poll the websocket
+            raw_message = self.ws.recv()
+        except Exception as err:  # pylint: disable=broad-except
+            # If the connection is closed, reconnect
+            self.log(
+                "Websocket connection closed. "
+                f"Reconnecting in {self.ws_params.retry_sleep} "
+                f"seconds... The following error occurred: {err}",
+                level="error",
+            )
+            time.sleep(self.ws_params.retry_sleep)
+            self.create_subscription()
+            return
+
+        try:
+            # Parse the message and emit to producers
+            message = json.loads(raw_message)
+            if self.ws_params.metadata is not None:
+                message = {
+                    "metadata": self.ws_params.metadata,
+                    "data": message,
+                }
+            for dataset, producer in self.producers.items():
+                if dataset != "logging":
+                    producer.produce(message)
```

Comment:
Similar to the comment above, but here it looks like you could give the websocket any number of producer topics that would have duplicate data from all messages. Would it be worth routing before that, or is the goal here to capture just the raw websocket data which could be processed later?

Reply:
i'm not a big fan of doing any sort of routing in this node. this node is intended to strictly speaking write raw data from the websocket to be processed downstream. i did this here to handle scenarios where the user sets more than one output dataset. since we are writing the same data and copying shouldn't be useful, i'll update this logic to validate that there is only one output dataset.